File: kernel/coc/rules/mythos.py
"""Cthulhu Mythos tracking and max-SAN coupling (Chapter 8). Ported from coc_mythos.py.

The kernel keeps `cm_value`, `max_san` and `current_san` on the party sheet, so the
persisted variants here mutate the sheet dict the caller then writes; there is no
separate investigator-state file."""

from __future__ import annotations

from typing import Any

FIRST_ENCOUNTER_GAIN = 5
SUBSEQUENT_ENCOUNTER_GAIN = 1
BASE_MAX_SAN = 99


def max_san_for(cm_value: int) -> int:
    """max_san = 99 - cm_value (p.167, F9). Floors at 0."""
    return max(0, BASE_MAX_SAN - int(cm_value))


def gain_mythos(investigator_state: dict[str, Any], *, amount: int | None = None,
                is_first: bool = False) -> dict[str, Any]:
    """Apply a Cthulhu Mythos gain: cm_value rises, max_san drops, current_san clamps."""
    cm_before = int(investigator_state.get("cm_value", 0))
    gain = (FIRST_ENCOUNTER_GAIN if is_first else SUBSEQUENT_ENCOUNTER_GAIN) if amount is None else int(amount)
    cm_after = cm_before + gain
    max_san_before = int(investigator_state.get("max_san", BASE_MAX_SAN - cm_before))
    max_san_after = max_san_for(cm_after)
    san_clamped = 0
    current_san = investigator_state.get("current_san")
    if current_san is not None:
        current_san = int(current_san)
        if current_san > max_san_after:
            san_clamped = current_san - max_san_after
            current_san = max_san_after
        investigator_state["current_san"] = current_san
    investigator_state["cm_value"] = cm_after
    investigator_state["max_san"] = max_san_after
    return {
        "event_type": "cthulhu_mythos_gain", "cm_before": cm_before, "cm_gain": gain, "cm_after": cm_after,
        "max_san_before": max_san_before, "max_san_after": max_san_after, "san_clamped": san_clamped,
        "is_first": is_first,
        "summary": (f"Cthulhu Mythos +{gain} ({cm_before}->{cm_after}); max SAN {max_san_before}->{max_san_after}"
                    + (f"; current SAN clamped by {san_clamped}" if san_clamped else "") + "."),
    }


def become_believer(investigator_state: dict[str, Any], *, source: str = "first_hand_encounter",
                    mythos_gain: int | None = None, is_first: bool = False) -> dict[str, Any]:
    """p.179: a first-hand encounter forces belief and costs SAN equal to current CM;
    a tome lets the investigator choose not to believe (CM still rises)."""
    if source not in ("first_hand_encounter", "tome"):
        raise ValueError(f"unsupported believer source: {source!r}")
    cm_before = int(investigator_state.get("cm_value", 0))
    san_before = investigator_state.get("current_san")
    san_before_int = int(san_before) if san_before is not None else None
    san_lost = 0
    permanently_insane = False
    if source == "first_hand_encounter" and san_before_int is not None:
        san_lost = cm_before
        san_before_int = max(0, san_before_int - san_lost)
        investigator_state["current_san"] = san_before_int
        permanently_insane = san_before_int == 0
    gain_event = gain_mythos(investigator_state, amount=mythos_gain, is_first=is_first)
    investigator_state["believer"] = True
    return {
        "event_type": "become_believer", "source": source, "cm_before": cm_before,
        "cm_after": gain_event["cm_after"], "san_lost": san_lost, "san_after": san_before_int,
        "permanently_insane": permanently_insane, "max_san_after": gain_event["max_san_after"],
        "is_first": is_first, "believer": True, "mythos_gain_event": gain_event,
        "rule_ref": "core.mythos.become_believer",
        "summary": (f"Became believer ({source}): CM {cm_before}->{gain_event['cm_after']}, "
                    + (f"SAN -{san_lost}" if san_lost else "no SAN lost (tome, chose not to believe)")
                    + (", permanent insanity" if permanently_insane else "") + "."),
    }

File: kernel-ts/magic/mythos.ts
/** Existing Cthulhu Mythos gain and believer state coupling, without new decisions. */
import { repr, type Row } from '../read/values.js';
import { valueError } from '../resolve/arithmetic.js';
import { sanityInt as int } from '../sanity/expression.js';

const field = (value: Row, key: string, fallback: any): any => Object.hasOwn(value, key) ? value[key] : fallback;
export const FIRST_ENCOUNTER_GAIN = 5;
export const SUBSEQUENT_ENCOUNTER_GAIN = 1;
export const BASE_MAX_SAN = 99;
export const maxSanFor = (cmValue: number): number => Math.max(0, BASE_MAX_SAN - int(cmValue));
export function gainMythos(state: Row, options: { amount?: number | null; isFirst?: boolean } = {}): Row {
    const before = int(field(state, 'cm_value', 0)), first = options.isFirst ?? false;
    const gain = options.amount == null ? first ? FIRST_ENCOUNTER_GAIN : SUBSEQUENT_ENCOUNTER_GAIN : int(options.amount), after = before + gain;
    const maxBefore = int(field(state, 'max_san', BASE_MAX_SAN - before)), maxAfter = maxSanFor(after);
    let clamped = 0;
    if (state.current_san != null) {
        let san = int(state.current_san);
        if (san > maxAfter) { clamped = san - maxAfter; san = maxAfter; }
        state.current_san = san;
    }
    state.cm_value = after; state.max_san = maxAfter;
    return { event_type: 'cthulhu_mythos_gain', cm_before: before, cm_gain: gain, cm_after: after, max_san_before: maxBefore, max_san_after: maxAfter,
        san_clamped: clamped, is_first: first,
        summary: `Cthulhu Mythos +${gain} (${before}->${after}); max SAN ${maxBefore}->${maxAfter}${clamped ? `; current SAN clamped by ${clamped}` : ''}.` };
}
export function becomeBeliever(state: Row, options: { source?: string; mythosGain?: number | null; isFirst?: boolean } = {}): Row {
    const source = options.source ?? 'first_hand_encounter', first = options.isFirst ?? false;
    if (!['first_hand_encounter', 'tome'].includes(source)) valueError(`unsupported believer source: ${repr(source)}`);
    const cm = int(field(state, 'cm_value', 0));
    let san = state.current_san == null ? null : int(state.current_san), lost = 0, permanentlyInsane = false;
    if (source === 'first_hand_encounter' && san !== null) {
        lost = cm; san = Math.max(0, san - lost); state.current_san = san; permanentlyInsane = san === 0;
    }
    const gain = gainMythos(state, { amount: options.mythosGain, isFirst: first }); state.believer = true;
    return { event_type: 'become_believer', source, cm_before: cm, cm_after: gain.cm_after, san_lost: lost, san_after: san,
        permanently_insane: permanentlyInsane, max_san_after: gain.max_san_after, is_first: first, believer: true, mythos_gain_event: gain,
        rule_ref: 'core.mythos.become_believer',
        summary: `Became believer (${source}): CM ${cm}->${gain.cm_after}, ${lost ? `SAN -${lost}` : 'no SAN lost (tome, chose not to believe)'}${permanentlyInsane ? ', permanent insanity' : ''}.` };
}
